topi/python/topi/cuda/rcnn/proposal.py

```diff
@@ -0,0 +1,356 @@
+
+    with ib.for_range(0, batch, for_type="unroll") as b:
+        start = b * num_bbox
+        for i in range(2):
+            bbox_id = tid * 2 + i
+            with ib.if_scope(bbox_id < num_bbox):
+                index_out[start + bbox_id] = bbox_id
+        with ib.for_range(0, num_bbox) as k:
+            offset = start + 2 * tid + (k % 2)
+            with ib.if_scope(
+                tvm.all(offset + 1 < num_bbox, p_data[offset] < p_data[offset + 1])):
+                temp_data[0] = p_data[offset]
+                p_data[offset] = p_data[offset + 1]
+                p_data[offset + 1] = temp_data[0]
+                temp_index[0] = index_out[offset]
+                index_out[offset] = index_out[offset + 1]
+                index_out[offset + 1] = temp_index[0]
+            ib.emit(tvm.make.Call(None, 'tvm_storage_sync',
+                                  tvm.convert(['shared']),
+                                  tvm.expr.Call.Intrinsic, None, 0))
+    return ib.get()
+
+
+def nms_ir(sorted_bbox_buf, out_buf, nms_threshold):
+    """Non-maximum supression.
+
+    Parameters
+    ----------
+    sorted_bbox_buf : tvm.schedule.Buffer
+        3-D with shape [batch, num_bbox, 5]. The last dimension is in format of
+        [w_start, h_start, w_end, h_end, score].
+
+    out_buf : tvm.schedule.Buffer
+        2-D with shape [batch, num_bbox]. Boolean mask of whether a bounding box should be removed.
+
+    nms_threshold : float
+        Non-maximum suppression threshold.
+
+    Returns
+    -------
+    stmt : Stmt
+        The result IR statement.
+    """
+    def calculate_overlap(out_tensor, box_a_idx, box_b_idx):
+        """Calculate overlap of two boxes.
+        """
+        w = tvm.max(0.0, tvm.min(out_tensor[box_a_idx + 2], out_tensor[box_b_idx + 2])
+                    - tvm.max(out_tensor[box_a_idx], out_tensor[box_b_idx]) + 1.0)
+        h = tvm.max(0.0, tvm.min(out_tensor[box_a_idx + 3], out_tensor[box_b_idx + 3])
+                    - tvm.max(out_tensor[box_a_idx + 1], out_tensor[box_b_idx + 1]) + 1.0)
+        i = w * h
+        u = (out_tensor[box_a_idx + 2] - out_tensor[box_a_idx] + 1.0) * \
+            (out_tensor[box_a_idx + 3] - out_tensor[box_a_idx + 1] + 1.0) + \
+            (out_tensor[box_b_idx + 2] - out_tensor[box_b_idx] + 1.0) * \
+            (out_tensor[box_b_idx + 3] - out_tensor[box_b_idx + 1] + 1.0) - i
+        return i / u
+
+    batch, num_bbox = get_const_tuple(out_buf.shape)
+    max_threads = int(math.sqrt(tvm.target.current_target(allow_none=False).max_num_threads))
+    tx = tvm.thread_axis("threadIdx.x")
+    bx = tvm.thread_axis("blockIdx.x")
+    ib = tvm.ir_builder.create()
+    p_data = ib.buffer_ptr(sorted_bbox_buf)
+    p_out = ib.buffer_ptr(out_buf)
+    nthread_tx = max_threads
+    nthread_bx = num_bbox // max_threads + 1
+    ib.scope_attr(tx, "thread_extent", nthread_tx)
+    ib.scope_attr(bx, "thread_extent", nthread_bx)
```

Comment:
@Laurawly In nmr_ir, sometimes valid bboxes are dropped due to conflicts. The only thing I do to fix this issue is binding bx to virtual threads. Could you take a look here?

Reply:
@vinx13 Your nms_ir looks very similar as mine. I haven't faced test cases which will drop bboxes. But one solution I have in mind is instead of parallelize around `i`, shall we parallelize around `l` and put `i` to for loop for serialized writing to p_out?

Reply:
@Laurawly I don't know how `l` can be parallelized since sequential dependency on `l` is required.

Reply:
@vinx13 I recently tested nms on Mali GPU and the data race occurred, when I changed the `blockIdx.x` to `vthread` and added synchronization it worked. But I was able to use `blockIdx.x` if I initialized the p_out to -1, but I still need the synchronization to make it work.